internal/pkg/agent/application/upgrade/marker_watcher.go

```diff
@@ -0,0 +1,123 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+type MarkerWatcher interface {
+	Watch() <-chan UpdateMarker
+	Run(ctx context.Context)
+	Close() error
+}
+
+type MarkerFileWatcher struct {
+	watcher        *fsnotify.Watcher
+	markerFilePath string
+
+	logger *logger.Logger
+
+	updateCh chan UpdateMarker
+}
+
+func newMarkerFileWatcher(upgradeMarkerFilePath string, logger *logger.Logger) (MarkerWatcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create upgrade marker watcher: %w", err)
+	}
+
+	// Watch the upgrade marker file's directory, not the file itself, so we
+	// notice the file even if it's deleted and recreated.
+	upgradeMarkerDirPath := filepath.Dir(upgradeMarkerFilePath)
+	err = watcher.Add(upgradeMarkerDirPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set watch on upgrade marker's directory [%s]: %w", upgradeMarkerDirPath, err)
+	}
+
+	logger = logger.Named("marker_file_watcher")
+
+	return &MarkerFileWatcher{
+		watcher:        watcher,
+		markerFilePath: upgradeMarkerFilePath,
+		logger:         logger,
+		updateCh:       make(chan UpdateMarker),
+	}, nil
+}
+
+func (mfw *MarkerFileWatcher) Watch() <-chan UpdateMarker {
+	return mfw.updateCh
+}
+
+func (mfw *MarkerFileWatcher) Run(ctx context.Context) {
+	// Do an initial read from the upgrade marker file, in case the file
+	// is already present before the watching starts.
+	doInitialRead := make(chan struct{}, 1)
+	doInitialRead <- struct{}{}
+
+	// Handle watching
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case err, ok := <-mfw.watcher.Errors:
+				if !ok { // Channel was closed (i.e. Watcher.Close() was called).
+					mfw.logger.Debug("fsnotify.Watcher's error channel was closed")
+					return
```

Comment:
shouldn't we close the watcher when returning? Or we are relying on the caller to always do it ?

Reply:
As of the current implementation, we're relying on the caller to always do it but this could all change depending on what I find after https://github.com/elastic/elastic-agent/pull/3694#discussion_r1389177469.

Reply:
After simplifying the `UpgradeMarkerWatcher` implementation in 8b9d67c797848ffc33c2c3fe39bf9751b2aded64, I was able to move the `fsnotify` watcher's initialization and closing into the `Run()` method in 49d16e94ef.  As such, the caller doesn't need to do anything any more.